OpenRA.Mods.Common/Traits/World/ActorMap.cs

```diff
@@ -616,6 +618,39 @@ public IEnumerable<Actor> ActorsInBox(WPos a, WPos b)
 				}
 			}
 		}
+
+		public bool IsBuildingBlocked(CPos cell)
+		{
+			return buildingBlocked.ContainsKey(cell);
+		}
+
+		public void AddBuildingBlocker(object blocker, IEnumerable<CPos> footprint)
+		{
+			// Take a read-only copy of the footprint cells
+			var cells = footprint.ToArray();
+			buildingBlockers.Add(blocker, cells);
+			foreach (var c in cells)
+			{
+				if (!buildingBlocked.TryGetValue(c, out var blockers))
+					blockers = buildingBlocked[c] = new List<object>();
+
+				blockers.Add(blocker);
+			}
+		}
+
+		public void RemoveBuildingBlocker(object blocker)
+		{
+			var footprint = buildingBlockers[blocker];
+			buildingBlockers.Remove(blocker);
+
+			foreach (var c in footprint)
+			{
+				var blockers = buildingBlocked[c];
+				blockers.Remove(blocker);
+				if (blockers.Count == 0)
```

Comment:
This is a bit weird. You are only removing it for real if nothing is left so that the removal is complete (all or nothing)? Why is this necessary?

Reply:
We are taking advantage of the fact that the total number of blocked cells on the map is small and sparse to use a dictionary of (blocked cell) &rarr; (list of blockers) instead of having to define an entire `CellLayer`. If the dictionary doesn't contain an entry for a cell we know immediately that cell is not blocked. When we remove a building we want to clean up the empty list so this remains true, instead of having to look up the list and then check if it is empty.